File: AssetManager/Helpers/OtherFunctions.vb
﻿Option Explicit On

Imports System.IO
Imports MyDialogLib

Module OtherFunctions
    Public stpw As New Stopwatch
    Private intTimerHits As Integer = 0

    Public Sub StartTimer()
        stpw.Stop()
        stpw.Reset()
        stpw.Start()
    End Sub

    Public Function StopTimer() As String
        stpw.Stop()
        intTimerHits += 1
        Dim Results As String = intTimerHits & "  Stopwatch: MS:" & stpw.ElapsedMilliseconds & " Ticks: " & stpw.ElapsedTicks
        Debug.Print(Results)
        Return Results
    End Function

    Public Sub Logger(Message As String)
        Dim MaxLogSizeKiloBytes As Short = 500
        Dim DateStamp As String = DateTime.Now.ToString
        Dim infoReader As FileInfo
        infoReader = My.Computer.FileSystem.GetFileInfo(strLogPath)
        If Not File.Exists(strLogPath) Then
            Directory.CreateDirectory(strAppDir)
            Using sw As StreamWriter = File.CreateText(strLogPath)
                sw.WriteLine(DateStamp & ": Log Created...")
                sw.WriteLine(DateStamp & ": " & Message)
            End Using
        Else
            If (infoReader.Length / 1000) < MaxLogSizeKiloBytes Then
                Using sw As StreamWriter = File.AppendText(strLogPath)
                    sw.WriteLine(DateStamp & ": " & Message)
                End Using
            Else
                If RotateLogs() Then
                    Using sw As StreamWriter = File.AppendText(strLogPath)
                        sw.WriteLine(DateStamp & ": " & Message)
                    End Using
                End If
            End If
        End If
    End Sub

    Private Function RotateLogs() As Boolean
        Try
            File.Copy(strLogPath, strLogPath + ".old", True)
            File.Delete(strLogPath)
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Public Sub EndProgram()
        GlobalSwitches.ProgramEnding = True
        Logger("Ending Program...")
        PurgeTempDir()
        Application.Exit()
    End Sub

    Public Sub PurgeTempDir()
        Try
            Directory.Delete(DownloadPath, True)
        Catch
        End Try
    End Sub

    Public Sub AdjustComboBoxWidth(ByVal sender As Object, ByVal e As EventArgs)
        Dim senderComboBox = DirectCast(sender, ComboBox)
        Dim width As Integer = senderComboBox.DropDownWidth
        Dim g As Graphics = senderComboBox.CreateGraphics()
        Dim font As Font = senderComboBox.Font
        Dim vertScrollBarWidth As Integer = If((senderComboBox.Items.Count > senderComboBox.MaxDropDownItems), SystemInformation.VerticalScrollBarWidth, 0)
        Dim newWidth As Integer
        For Each s As String In DirectCast(sender, ComboBox).Items
            newWidth = CInt(g.MeasureString(s, font).Width) + vertScrollBarWidth
            If width < newWidth Then
                width = newWidth
            End If
        Next
        senderComboBox.DropDownWidth = width
    End Sub

    Public Function NotePreview(Note As String, Optional CharLimit As Integer = 50) As String
        If Note <> "" Then
            Return Strings.Left(Note, CharLimit) & IIf(Len(Note) > CharLimit, "...", "").ToString
        Else
            Return ""
        End If
    End Function

    Public Function Message(ByVal Prompt As String, Optional ByVal Buttons As Integer = vbOKOnly + vbInformation, Optional ByVal Title As String = Nothing, Optional ByVal ParentFrm As Form = Nothing) As MsgBoxResult
        SetWaitCursor(False, ParentFrm)
        Dim NewMessage As New AdvancedDialog(ParentFrm)
        Return NewMessage.DialogMessage(Prompt, Buttons, Title, ParentFrm)
    End Function

    Public Function OKToEnd() As Boolean
        If GlobalSwitches.BuildingCache Then
            Message("Still building DB Cache. Please wait and try again.", vbOKOnly + vbInformation, "Critical Function Running")
            Return False
        End If
        If GKUpdaterForm.Visible AndAlso Not GKUpdaterForm.OkToClose() Then Return False
        Return True
    End Function

    Public Sub SetWaitCursor(Waiting As Boolean, Optional parentForm As Form = Nothing)
        If parentForm Is Nothing Then
            Application.UseWaitCursor = Waiting
            Application.DoEvents()
        Else
            If Waiting Then
                parentForm.Cursor = Cursors.AppStarting
            Else
                parentForm.Cursor = Cursors.Default
            End If
            parentForm.Update()
        End If

    End Sub

End Module

File: AssetManager/Helpers/OtherFunctions.cs
using AdvancedDialog;
using System;
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using AssetManager.Helpers.Watchdog;

namespace AssetManager.Helpers
{
    internal static class OtherFunctions
    {
        private static RichTextBox rtfBox = new RichTextBox();
        public static Stopwatch stpw = new Stopwatch();

        private static int intTimerHits = 0;

        public static void StartTimer()
        {
            stpw.Stop();
            stpw.Reset();
            stpw.Start();
        }

        public static string StopTimer()
        {
            stpw.Stop();
            intTimerHits += 1;
            string Results = intTimerHits + "  Stopwatch: MS:" + stpw.ElapsedMilliseconds + " Ticks: " + stpw.ElapsedTicks;
            Console.WriteLine(Results);
            return Results;
        }

        public static string ElapTime()
        {
            string results = intTimerHits + "  Elapsed: MS:" + stpw.ElapsedMilliseconds + " Ticks: " + stpw.ElapsedTicks;
            Console.WriteLine(results);
            return results;
        }

        public static void EndProgram(bool forceEnd = false)
        {
            GlobalSwitches.ProgramEnding = true;
            Logging.Logger("Ending Program...");
            WatchdogInstance.Watchdog.Dispose();
            PurgeTempDir();

            // Forcefully kill the process. Used when unhandled or critical exceptions occur.
            if (forceEnd)
            {
                Environment.Exit(-1);
            }
        }

        public static void PurgeTempDir()
        {
            try
            {
                if (Directory.Exists(Paths.DownloadPath))
                {
                    Directory.Delete(Paths.DownloadPath, true);
                }
            }
            catch
            {
                // Sometimes a file will be in use, so we don't care if this fails.
            }
        }

        public static string NotePreview(string text, int maxChars = 50)
        {
            if (!string.IsNullOrEmpty(text))
            {
                if (text.Length > maxChars)
                {
                    return text.Substring(0, maxChars) + "...";
                }
                else
                {
                    return text;
                }
            }
            else
            {
                return string.Empty;
            }
        }

        public static DialogResult Message(string prompt, MessageBoxButtons button = MessageBoxButtons.OK, MessageBoxIcon icon = MessageBoxIcon.Information, string title = null, Form parentForm = null)
        {
            SetWaitCursor(false, parentForm);
            using (var newMessage = new Dialog(parentForm))
            {
                return newMessage.DialogMessage(prompt, button, icon, title, parentForm);
            }
        }

        public static bool OKToEnd()
        {
            if (GlobalSwitches.BuildingCache)
            {
                Message("Still building DB Cache. Please wait and try again.", MessageBoxButtons.OK, MessageBoxIcon.Information, "Critical Function Running");
                return false;
            }
            return true;
        }

        private static void SetWaitCursor(bool waiting, Form parentForm)
        {
            if (parentForm == null)
            {
                Application.UseWaitCursor = waiting;
                return;
            }
            
            if (parentForm.InvokeRequired)
            {
                var del = new Action(() => SetWaitCursor(waiting, parentForm));
                parentForm.BeginInvoke(del);
            }
            else
            {
                if (waiting)
                {
                    parentForm.UseWaitCursor = true;
                    Cursor.Current = Cursors.WaitCursor;
                }
                else
                {
                    parentForm.UseWaitCursor = false;
                    Cursor.Current = Cursors.Default;
                }
            }
        }

        public static string RTFToPlainText(string rtfText)
        {
            try
            {
                if (rtfText.StartsWith("{\\rtf"))
                {
                    rtfBox.Rtf = rtfText;
                    return rtfBox.Text;
                }
                else
                {
                    return rtfText;
                }
            }
            catch (ArgumentException)
            {
                //If we get an argument error, that means the text is not RTF so we return the plain text.
                return rtfText;
            }
        }
    }
}